File: src/test/java/kitchenpos/order/tobe/eatinorder/domain/OrderTableTest.java
package kitchenpos.order.tobe.eatinorder.domain;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import kitchenpos.Fixtures;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

class OrderTableTest {

    @DisplayName("주문 테이블을 초기화한다.")
    @Test
    void testChangeEmptyTable() {
        // given
        OrderTable orderTable = Fixtures.orderTable(true, 3);

        // when
        orderTable.changeEmptyTable();

        // then
        assertThat(orderTable.getNumberOfGuests()).isZero();
        assertThat(orderTable.isOccupied()).isFalse();
    }

    @DisplayName("주문 테이블을 사용중인 테이블로 변경한다.")
    @Test
    void testChangeInUseTable() {
        // given
        OrderTable orderTable = Fixtures.orderTable(false, 0);

        // when
        orderTable.changeInUseTable();

        // then
        assertThat(orderTable.getNumberOfGuests()).isZero();
        assertThat(orderTable.isOccupied()).isTrue();
    }

    @DisplayName("주문 테이블의 손님 수를 변경한다.")
    @Test
    void testChangeNumberOfGuests() {
        // given
        OrderTable orderTable = Fixtures.orderTable(true, 2);
        int expectedNumberOfGuests = 3;

        // when
        orderTable.changeNumberOfGuests(expectedNumberOfGuests);

        // then
        assertThat(orderTable.getNumberOfGuests()).isEqualTo(expectedNumberOfGuests);
        assertThat(orderTable.isOccupied()).isTrue();
    }

    @DisplayName("빈 주문 테이블은 손님 수를 변경할 수 없다")
    @Test
    void testChangeNumberOfGuestsIfEmptyOrderTable() {
        // given
        OrderTable orderTable = Fixtures.emptyOrderTable();
        int expectedNumberOfGuests = 3;

        // when // then
        assertThatThrownBy(() -> orderTable.changeNumberOfGuests(expectedNumberOfGuests))
            .isExactlyInstanceOf(IllegalStateException.class);
    }

    @DisplayName("주문 테이블이 비어있는지 확인한다")
    @ParameterizedTest
    @CsvSource(value = {"false;true", "true;false"}, delimiter = ';')
    void testIsEmpty(boolean occupied, boolean expected) {
        // given
        var orderTable = Fixtures.orderTable(occupied, 0);

        // when
        boolean actual = orderTable.isEmpty();

        // then
        assertThat(actual).isEqualTo(expected);
    }
}

File: src/test/kotlin/kitchenpos/order/tobe/eatinorder/domain/OrderTableTest.kt
package kitchenpos.order.tobe.eatinorder.domain

import java.util.*
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTable
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTableEmptyService
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTableName
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTableOccupancy
import kitchenpos.order.tobe.eatinorder.domain.ordertable.OrderTableStatus
import kitchenpos.utils.Fixtures
import org.assertj.core.api.Assertions.assertThatIllegalStateException
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

class OrderTableTest {
    @Test
    @DisplayName("OrderTable을 생성한다.")
    fun create() {
        // when
        val orderTable = OrderTable(
            id = UUID.randomUUID(),
            orderTableName = OrderTableName("테이블1"),
            orderTableOccupancy = OrderTableOccupancy(0, OrderTableStatus.EMPTY)
        )

        // then
        assertAll(
            { assertNotNull(orderTable.id) },
            { assertEquals(orderTable.orderTableName.name, "테이블1") },
            { assertEquals(orderTable.orderTableOccupancy.numberOfGuests, 0) },
            { assertEquals(orderTable.orderTableOccupancy.status, OrderTableStatus.EMPTY) }
        )
    }

    @Test
    @DisplayName("OccupiedTable의 numberOfGuest를 변경한다.")
    fun changeNumberOfGuestOccupiedTable() {
        // given
        val orderTable = Fixtures.orderTable(orderTableOccupancy = OrderTableOccupancy(0, OrderTableStatus.OCCUPIED))

        // when
        orderTable.changeNumberOfGuest(4)

        // then
        assertEquals(orderTable.orderTableOccupancy.numberOfGuests, 4)
    }


    @Test
    @DisplayName("EmptyTable을 OccupiedTable로 변경한다")
    fun changeOccupiedTable() {
        // given
        val orderTable = Fixtures.orderTable(orderTableOccupancy = OrderTableOccupancy(0, OrderTableStatus.EMPTY))

        // when
        orderTable.occupied()

        // then
        assertAll(
            { assertEquals(orderTable.orderTableOccupancy.numberOfGuests, 0) },
            { assertEquals(orderTable.orderTableOccupancy.status, OrderTableStatus.OCCUPIED) }
        )
    }

    @Test
    @DisplayName("OccupiedTable을 EmptyTable로 변경한다")
    fun changeEmptyTable() {
        // given
        val orderTable = Fixtures.orderTable(orderTableOccupancy = OrderTableOccupancy(4, OrderTableStatus.OCCUPIED))

        // when
        orderTable.empty(object : OrderTableEmptyService {
            override fun canEmpty(orderTable: OrderTable): Boolean {
                return true
            }
        })

        // then
        assertAll(
            { assertEquals(orderTable.orderTableOccupancy.numberOfGuests, 0) },
            { assertEquals(orderTable.orderTableOccupancy.status, OrderTableStatus.EMPTY) }
        )
    }

    @Test
    @DisplayName("OccupiedTable을 EmptyTable로 변경할 수 없으면 에러가 발생한다")
    fun changeEmptyTableFail() {
        // given
        val orderTable = Fixtures.orderTable(orderTableOccupancy = OrderTableOccupancy(4, OrderTableStatus.OCCUPIED))

        // when
        assertThatIllegalStateException().isThrownBy {
            orderTable.empty(object : OrderTableEmptyService {
                override fun canEmpty(orderTable: OrderTable): Boolean {
                    return false
                }
            })
        }
    }

    @Test
    @DisplayName("OccupiedTable을 EmptyTable로 변경할 수 있으면 변경한다")
    fun changeEmptyTableIfPossible() {
        // given
        val orderTable = Fixtures.orderTable(orderTableOccupancy = OrderTableOccupancy(4, OrderTableStatus.OCCUPIED))
        val orderTable2 = Fixtures.orderTable(orderTableOccupancy = OrderTableOccupancy(4, OrderTableStatus.OCCUPIED))

        // when
        orderTable.emptyIfPossible(object : OrderTableEmptyService {
            override fun canEmpty(orderTable: OrderTable): Boolean {
                return true
            }
        })
        orderTable2.emptyIfPossible(object : OrderTableEmptyService {
            override fun canEmpty(orderTable: OrderTable): Boolean {
                return false
            }
        })

        // then
        assertAll(
            { assertEquals(orderTable.orderTableOccupancy.numberOfGuests, 0) },
            { assertEquals(orderTable.orderTableOccupancy.status, OrderTableStatus.EMPTY) },
            { assertEquals(orderTable2.orderTableOccupancy.numberOfGuests, 4) },
            { assertEquals(orderTable2.orderTableOccupancy.status, OrderTableStatus.OCCUPIED) }
        )
    }
}
